### DISCUSSION TITLE
FR: User-defined default pasted image format

### DISCUSSION THREAD

**Discussion by introt** [zim-desktop-wiki/zim-desktop-wiki#1937] [Ideas]
**Is your feature request related to a problem? Please describe.**
Pasted images are (almost) always `png`s, resulting in unnecessarily large files.

**Describe the solution you'd like**
Being able to set the default image format to `jpeg` (supported formats [here](https://lazka.github.io/pgi-docs/GdkPixbuf-2.0/classes/Pixbuf.html#GdkPixbuf.Pixbuf.savev)).

**Describe alternatives you've considered**
Populating the right-click menu with `Paste As...` > `Image (jpeg)`

**Additional context**
Current implementation here:
https://github.com/zim-desktop-wiki/zim-desktop-wiki/blob/3e9064f2dda99b66e26de8adbf566b6fef1b32c5/zim/gui/clipboard.py#L179-L184

**Comment by introt** [comment]
#1236 introduced Pillow for viewing additional image formats; it could be used here as well to provide support for formats other than _“jpeg”, “png”, “tiff”, “ico” or “bmp”_.

**Reply by hsandt** [reply]
Cool! I started using avif recently (mostly at quality 90 which seems a good compromise, and quality 100 makes it bigger than a lossless webp anyway so wasn't useful), unfortunately, it's not well supported in general (e.g. on Linux Ubuntu, I must visualize it with gThumb while default Image Viewer (eyes of gnome) can open webp), and Pillow doesn't support it either (I can see that dragging an avif to Zim fails to preview it), however there is a plugin https://pypi.org/project/pillow-avif-plugin/ so it may be possible to support in the future.

In this case, Paste As... > Image (avif quality 90) would be nice (and also lossless webp when I need something lossless but smaller than png). As a user, I'd like to verify if the image I'm copying was not already a jpg or webp in the first place though, so it would be great if there was an indication of this in the contextual menu, so I decide if it needs more size reduction or not.

**Comment by introt** [comment]
Could also have an "auto" option which chooses the best format according to smallest file size, as well as a toggle for forcing the selected format regardless of the detected one.

**Reply by dehesselle** [reply]
If smallest file size is the intention, name it `smallest` or something. Mainly because I don't like `auto` being possibly a lossy format.

**Reply by introt** [reply]
Good call, thanks. `Default format: [smallest]` + `[x] Always use default format` sounds like a good combo to me - `auto` being the opposite of selecting the latter, and (only) as lossy as the source[^lossy].

[^lossy]: At least that seems to be the intention of the existing code - on Linux, I've only ever got pngs; you can try copying and pasting https://upload.wikimedia.org/wikipedia/commons/0/0e/Felis_silvestris_silvestris.jpg to verify.

**Comment by introt** [comment]
Related feature request: setting a default display size for pasted images, with a toggle for resizing the actual images before saving. Could also have some preset sizes available to choose from in the Paste As menu

See also: #1423

Note: While can in practice make images fit in the viewport, this is a different feature aimed at setting dimensions for exports and/or saving disk space and is thus not related to #324 where the solution is implementing a new widget.
